File: FTPUploader/VB/FtpInfo.vb
Imports System.IO

Public Class FtpInfo

	Private Shared cred As FtpInfo, options As New List(Of FtpInfo)

	Public Shared ReadOnly Property Selected() As FtpInfo
		Get
			Return cred
		End Get
	End Property

	Private Const CONFIGFILE As String = "FTPSync.ini"
	Private Const [Default] = "Default=c:\web|ftp://ftp.website.com/www/|user|"

#Region "Options - Loading / Selecting"

	Shared Sub New()
		ReadOptions()

		If Program.SilentMode Then Exit Sub
		cred = Nothing
		If HasMultipleOptions Then SelectOption()
	End Sub

	Private Shared Sub ReadOptions()
		Dim lines() As String
		If File.Exists(CONFIGFILE) Then
			lines = File.ReadAllLines(CONFIGFILE)
		Else
			lines = New String() {[Default]}
		End If

		options.Clear()
		For Each line In lines
			Dim nv = line.Split("=", 2, StringSplitOptions.None)
			cred = New FtpInfo(nv(1))
			cred.Name = nv(0)
			options.Add(cred)
		Next
	End Sub

	Public Shared ReadOnly Property HasMultipleOptions() As Boolean
		Get
			Return options.Count > 1
		End Get
	End Property

	'Consider making separate method as in IViewer
	Public Shared Sub SelectOption()
		Dim frm As New Form()
		frm.Height = 150 : frm.Width = 450 : frm.Text = "Select FTP Connection / Folder" : frm.StartPosition = FormStartPosition.CenterScreen : frm.KeyPreview = True
		AddHandler frm.KeyDown, AddressOf frm_KeyDown
		Dim lst As New ListBox
		lst.Items.AddRange(options.ToArray)
		If cred Is Nothing Then lst.SelectedIndex = 0 Else lst.SelectedItem = cred
		lst.Dock = DockStyle.Fill
		frm.Controls.Add(lst)
		frm.ShowDialog()
		cred = CType(lst.SelectedItem, FtpInfo)
		'MessageBox.Show("Chosen " & Selected.ToString)
	End Sub

	Public Shared Sub SetSelected(ByVal name As String)
		For Each opt In FtpInfo.options
			If opt.Name = name Then cred = opt : Exit For
		Next
	End Sub

	Private Shared Sub frm_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs)
		If e.KeyCode = Keys.F5 Then
			ReadOptions()
			Dim lst As ListBox = CType(CType(sender, Form).Controls(0), ListBox)
			lst.Items.Clear()
			lst.Items.AddRange(options.ToArray)
			If cred Is Nothing Then lst.SelectedIndex = 0 Else lst.SelectedItem = cred
		End If
		If e.KeyCode = Keys.Return Then CType(sender, Form).Close()
	End Sub

#End Region

#Region "Properties / Instance Members"

	Private Sub New(ByVal val As String)
		Dim vals = val.Split("|")
		If vals.Length > 0 Then Value(Names.LocalFolder) = vals(0)
		If vals.Length > 1 Then Value(Names.FtpFolder) = vals(1)
		If vals.Length > 2 Then Value(Names.FtpUsername) = vals(2)
		If vals.Length > 3 Then Value(Names.FtpPassword) = vals(3)
	End Sub

	Public Name As String
	Dim _localFolder As String
	Dim _ftpFolder As String
	Dim _username As String
	Dim _password As String

	Public Property LocalFolder() As String
		Get
			Return _localFolder
		End Get
		Set(ByVal value As String)
			_localFolder = value
		End Set
	End Property

	Public Property FtpFolder() As String
		Get
			Return _ftpFolder
		End Get
		Set(ByVal value As String)
			_ftpFolder = value
		End Set
	End Property

	Public Property Username() As String
		Get
			Return _username
		End Get
		Set(ByVal value As String)
			_username = value
		End Set
	End Property

	Public Property Password() As String
		Get
			Return _password
		End Get
		Set(ByVal value As String)
			_password = value
		End Set
	End Property

	Public Property Value(ByVal name As Names)
		Get
			Select Case name
				Case Names.FtpFolder
					Return Me.FtpFolder
				Case Names.FtpPassword
					Return Me.Password
				Case Names.FtpUsername
					Return Me.Username
				Case Names.LocalFolder
					Return Me.LocalFolder
				Case Else
					Throw New NotSupportedException
			End Select
		End Get
		Set(ByVal value)
			Select Case name
				Case Names.FtpFolder
					Me.FtpFolder = value
				Case Names.FtpPassword
					Me.Password = value
				Case Names.FtpUsername
					Me.Username = value
				Case Names.LocalFolder
					Me.LocalFolder = value
				Case Else
					Throw New NotSupportedException
			End Select
		End Set
	End Property

	Public Overrides Function Equals(ByVal obj As Object) As Boolean
		Return Name.Equals(CType(obj, FtpInfo).Name)
	End Function

	Public Overrides Function GetHashCode() As Integer
		Return Name.GetHashCode()
	End Function

	Public Overrides Function ToString() As String
		Return String.Format("{0} ({1} on {2} @ {3})", Name, Username, FtpFolder, LocalFolder)
	End Function

#End Region

	Public Shared Function GetKey(ByVal key As Names) As String
		GetKey = Selected.Value(key)
		Dim cVal As String = GetKey

		'If String.IsNullOrEmpty(cVal) Then
		'dont show a textbox for FTP user / password if the ftp is null
		Dim ftpFor As String = IIf(key = Names.FtpUsername Or key = Names.FtpPassword, " for " + FtpInfo.Selected.FtpFolder, String.Empty)
		cVal = InputBox(key.ToString + ftpFor, "App Settings", cVal)
		'End If

		If cVal <> GetKey And String.IsNullOrEmpty(cVal) = False Then
			GetKey = cVal
			'FtpInfo.Selected.Value(key) = 
			'config(Array.IndexOf(configNames, cKey)) = GetKey
			'If File.Exists(CONFIGFILE) Then File.Delete(CONFIGFILE)
			'File.WriteAllText(CONFIGFILE, String.Join("|", config))
		End If
	End Function


End Class

File: FTPUploader/FtpInfo.cs
using Microsoft.VisualBasic;
using System;
using System.Collections.Generic;
using System.IO;
using System.Windows.Forms;

namespace AmadeusWeb.SmartSiteUploader
{
	public class FtpInfo
	{
		public static readonly string ConfigFile = StoreHelper.DataFolder + "FTPUploader.ini";

		private static readonly string Default = @"Amadeus All=D:\AmadeusWeb\all\|ftp://ftp.amadeusweb.com/all/|amadeus|";

		#region "Static ctor / Properties"

		static FtpInfo()
		{
			Options = new List<FtpInfo>();
			ReadOptions();

			if (Program.SilentMode)
				return;
			Selected = null;
			if (HasMultipleOptions)
			{
				SelectOption();
			}
			else
			{
				if (File.Exists(ConfigFile) == false)
					File.WriteAllText(ConfigFile, Default + IOHelper.Encode("passwd"));

				IOHelper.RunInNotepad(ConfigFile);
			}
		}

		public FtpInfo(string line)
		{
			var nv = line.Split("=".ToCharArray(), 2, StringSplitOptions.None);
			Name = nv[0];
			var val = nv[1];

			var vals = val.Split('|');
			if (vals.Length > 0)
				LocalFolder = vals[0];
			if (vals.Length > 1)
				FtpFolder = vals[1];
			if (vals.Length > 2)
				Username = vals[2];
			if (vals.Length > 3)
				Password = IOHelper.Decode(vals[3]);
		}

		public static List<FtpInfo> Options { get; private set; }

		public static FtpInfo Selected { get; private set; }

		public static bool HasMultipleOptions
		{
			get { return Options.Count > 1; }
		}

		#endregion

		#region Instance Properties

		public string Name { get; private set; }

		public string FtpFolder { get; set; }

		public string Password { get; set; }

		public string Username { get; set; }

		public string LocalFolder { get; set; }

		#endregion

		#region Static Methods

		public static void ReadOptions()
		{
			string[] lines = null;
			if (File.Exists(ConfigFile))
			{
				lines = File.ReadAllLines(ConfigFile);
			}
			else
			{
				lines = new string[] { Default, Default.Replace("Amadeus Main", "Your Project") };
			}

			Options.Clear();
			foreach (var line in lines)
			{
				var item = new FtpInfo(line);
				Options.Add(item);
			}
		}

		public static List<FtpInfo> Clone()
		{
			return Clone(Options);
		}

		public static void Save(List<FtpInfo> list)
		{
			Options = Clone(list);
            var lines = new List<string>();
            foreach (var line in list)
                lines.Add(line.ToLine());
			File.WriteAllLines(ConfigFile, lines.ToArray());
		}

		// Consider making separate method as in IViewer
		public static void SelectOption()
		{
			var frm = new Select(Selected);
			if (frm.ShowDialog() == DialogResult.Cancel) return;
			Selected = frm.Selected;
		}

		public static void SetSelected(string name)
		{
			foreach (var opt in FtpInfo.Options)
			{
				if (opt.Name == name)
				{
					Selected = opt;
					break; // TODO: might not be correct. Was : Exit For
				}
			}
		}

		public bool Matches(string what)
		{
			return Name.IndexOf(what, StringComparison.OrdinalIgnoreCase) != -1;
		}

		public string GetKey(Names key)
		{
			var retval = GetValNow(key);
			var val = retval;

			// dont show a textbox for FTP user / password if the ftp is null
			var ftpFor = key == Names.FtpUsername | key == Names.FtpPassword ? " for " + FtpInfo.Selected.FtpFolder : string.Empty;
			val = Interaction.InputBox(key.ToString() + ftpFor, "App Settings", val);

			if (val != retval & string.IsNullOrEmpty(val) == false)
			{
				retval = val;
			}

			return retval;
		}

		#endregion

		#region "Instance Methods"

		public override bool Equals(object obj)
		{
			if (obj == null || obj.Equals(DBNull.Value)) return false;
			return Name.Equals(((FtpInfo)obj).Name);
		}

		public override int GetHashCode()
		{
			return Name.GetHashCode();
		}

		public string ToLine()
		{
			// Default=c:\\web|ftp://ftp.website.com/www/|user|
			return string.Format("{0}={1}|{2}|{3}|{4}", Name, LocalFolder, FtpFolder, Username, IOHelper.Encode(Password));
		}

		public override string ToString()
		{
			return string.Format("{0} ({1} on {2} @ {3})", Name, Username, FtpFolder, LocalFolder);
		}

		private static List<FtpInfo> Clone(List<FtpInfo> list)
		{
            var r = new List<FtpInfo>();
            foreach (var item in list)
                r.Add((FtpInfo)item.MemberwiseClone());
            return r;
		}

		private string GetValNow(Names name)
		{
			switch (name)
			{
				case Names.FtpFolder:
					return this.FtpFolder;
				case Names.FtpPassword:
					return this.Password;
				case Names.FtpUsername:
					return this.Username;
				case Names.LocalFolder:
					return this.LocalFolder;
				default:
					throw new NotSupportedException();
			}
		}

		#endregion
	}
}